lib/ts/xmlhttprequest.ts

```diff
@@ -0,0 +1,429 @@
+/* Copyright (c) 2020, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { shouldDoInterceptionBasedOnUrl } from "./utils";
+import AuthHttpRequestFetch, {
+    AntiCsrfToken,
+    getIdRefreshToken,
+    setIdRefreshToken,
+    FrontToken,
+    onUnauthorisedResponse,
+    IdRefreshTokenType
+} from "./fetch";
+import { logDebugMessage } from "./logger";
+import WindowHandlerReference from "./utils/windowHandler";
+import { PROCESS_STATE, ProcessState } from "./processState";
+
+type XMLHttpRequestType = typeof XMLHttpRequest.prototype & { [key: string]: any };
+
+export function addInterceptorsToXMLHttpRequest() {
+    // create XMLHttpRequest proxy object
+    let oldXMLHttpRequest = XMLHttpRequest;
+
+    // define constructor for my proxy object
+    XMLHttpRequest = function(this: XMLHttpRequestType) {
+        let actual: XMLHttpRequestType = new oldXMLHttpRequest();
+
+        let self = this;
+        let listOfFunctionCallsInProxy: { (xhr: XMLHttpRequestType): void }[] = [];
+
+        let requestHeaders: { name: string; value: string }[] = [];
+        // let method: string = "";
+        let url: string | URL = "";
+        let doNotDoInterception = false;
+        let preRequestIdToken: IdRefreshTokenType | undefined = undefined;
+        let customGetterValues: { [key: string]: any } = {};
+
+        // we do not provide onerror cause that is fired only on
+        // network level failures and nothing else. If a status code is > 400,
+        // then onload and onreadystatechange are called.
+
+        self.onload = null;
+        self.onreadystatechange = null;
+        self.onloadend = null;
+
```

Comment:
how can we handle cases where a user is downloading a file which is guarded with verifySession?